modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/services/SentStateChannelBinaryTrackingService.scala

```diff
@@ -0,0 +1,88 @@
+package org.tessellation.currency.l0.snapshot.services
+
+import cats.effect.{Async, Ref}
+import cats.syntax.all._
+
+import org.tessellation.currency.l0.node.IdentifierStorage
+import org.tessellation.schema.GlobalIncrementalSnapshot
+import org.tessellation.security.signature.Signed
+import org.tessellation.security.{Hashed, Hasher}
+import org.tessellation.statechannel.StateChannelSnapshotBinary
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.numeric.{NonNegInt, PosInt}
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+trait SentStateChannelBinaryTrackingService[F[_]] {
+  def setPending(binary: Signed[StateChannelSnapshotBinary])(implicit hasher: Hasher[F]): F[Unit]
+  def getRetriable: F[List[Signed[StateChannelSnapshotBinary]]]
+  def updateByGlobalSnapshot(globalSnapshot: GlobalIncrementalSnapshot)(implicit hasher: Hasher[F]): F[Unit]
+}
+
+object SentStateChannelBinaryTrackingService {
+  private val retryOrdinalDelay: PosInt = 3
+
+  def make[F[_]: Async](
+    identifierStorage: IdentifierStorage[F]
+  ): F[SentStateChannelBinaryTrackingService[F]] =
+    Ref
+      .of[F, List[(Hashed[StateChannelSnapshotBinary], NonNegInt)]](
+        List.empty
+      )
+      .map(make[F](_, identifierStorage))
+
+  def make[F[_]: Async](
+    pendingR: Ref[F, List[(Hashed[StateChannelSnapshotBinary], NonNegInt)]],
+    identifierStorage: IdentifierStorage[F]
+  ): SentStateChannelBinaryTrackingService[F] =
+    new SentStateChannelBinaryTrackingService[F] {
+
+      private val logger = Slf4jLogger.getLogger
+
+      private def updateByGlobalSnapshotBinaries(
+        binaries: List[Signed[StateChannelSnapshotBinary]]
+      )(implicit hasher: Hasher[F]): F[Unit] = {
+        val confirmedHashes = binaries.traverse(_.toHashed).map(_.flatMap(b => List(b.hash, b.lastSnapshotHash)))
```

Comment:
Would it be more efficient if this were a `Set` since we're later doing `contains` in a loop?
Also, is there a chance that some hashes will be duplicated? E.g. `b.hash` in one binary is the `b.lastSnapshotHash` in another?

Reply:
I guess that PR isn't rebased on `develop` @marcinwadon because the service you are commenting @sdavidp doesn't exist anymore on `develop` and it already uses `Set`:

`confirmedHashesInGlobalSnapshot <- binaries.traverse(_.toHashed).map(_.map(_.hash)).map(_.toSet)`.

>Also, is there a chance that some hashes will be duplicated? E.g. b.hash in one binary is the b.lastSnapshotHash in another?

Yes, for example:
```
List(
   Binary(hash: "A", lastSnapshotHash: "000"),
   Binary(hash: "B", lastSnapshotHash: "A")
)
```

gives `List("A", "000" "B", "A")` which makes it even worse without `Set`.